proposed/http-client/http-client.md

```diff
@@ -21,9 +21,9 @@ implementations. This would make libraries more stable since the reduced number
 dependencies and the likelihood to get in version conflicts is reduced.
 
 The second goal is that all HTTP clients should follow the [Liskov substitutions principle][Liskov].
-This means that all clients should act the same when sending a request. By default a HTTP client
-should not follow redirect nor throw exceptions on HTTP responses with status 4xx or 5xx.
-
+This means that all clients should act the same when sending a request. A consuming
+library SHOULD expect a HTTP client to NOT follow redirects NOR throw exceptions based
+on HTTP response statuses.
```

Comment:
> i am not happy to open this can of worms

Me neither, but...

> reusable libraries would then be littered with configuration to each call where it should be configuration of the client behaviour

I strongly disagree that things like following redirects or (especially) throwing errors "should be" configuration of the client. In my opinion, at least these two options in particular, describe *not* how the client behaves, but how the individual request is processed.

As we've seen and discussed above, client behavior, when described as the state of the client, breaks SRP and effectively ends up forcing every consumer to create their own instance.

I think that's completely unnecessary. Any piece of consumer code, such as a REST client, that needs a particular consistent behavior from this method, can simply wrap the call in an internal method and supply the required options - which they would likely do anyway, to avoid manually constructing requests everywhere, adding authentication headers, etc. for example:

```php
class SomeRestClient 
{
    public function __construct(HttpClient $client) {
        $this->client = $client;
    }
    private function request(RequestInterface $request) {
        return $this->client->sendRequest($request, HttpClient::CHECK_STATUS);
    }
}
```

In other cases, where a consumer is performing only a single request, it doesn't really matter how these options get configured - it's the same workload, and this makes the expected behavior clearly visible at the call site, rather than needing to read through bootstrapping code elsewhere.

If you insist on request behavior options as state in the client, you'd need a client-factory interface - but it leads to the same end result, just more clutter and less transparency, so I'd be :-1: on that.

Yes, having a set of defined options *does* imply that each client implementation must be capable of doing certain things - isn't that the whole point of standardizing this in the first place though?

I'm not suggesting we brainstorm for an endless variety of "nice to have" options we can add, but I think, from the discussion above, it's pretty clear that there are common, real use-cases for the two options we've discussed here.

The client wouldn't be very useful without these, as indicated by your suggestion that decorators/wrappers could provide this behavior - but then, in my opinion, we're really talking about a client abstraction that's too low-level to be of any real-world use. If the abstraction provides only a low-level client, and every use-case needs some proprietary "client wrapper" to implement the major behavior, what have we won, really? How is a non-shared dependency on "client wrappers" any better than a non-shared dependency on the client itself?

For a client abstraction to succeed, it has to be useful in real use-cases - otherwise the whole thing seems like a waste of time.